File: apps/mobile/lib/services/notification_service.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:flutter_timezone/flutter_timezone.dart';
import 'package:timezone/data/latest_all.dart' as tz_data;
import 'package:timezone/timezone.dart' as tz;
import '../i18n/app_strings.dart';
import '../models/auth.dart';

/// Thin wrapper around `flutter_local_notifications` so the screen code only
/// has to call [reschedule] with the latest preferences. Re-entrant — re-runs
/// always start from a clean slate (cancelAll → schedule).
///
/// PLATFORM SETUP — see NOTIFICATIONS_SETUP.md for iOS Info.plist + Android
/// AndroidManifest.xml requirements. The first call to [requestPermission]
/// triggers the system prompt.
class NotificationService {
  NotificationService._();
  static final NotificationService instance = NotificationService._();

  final FlutterLocalNotificationsPlugin _plugin =
      FlutterLocalNotificationsPlugin();
  bool _initialised = false;

  /// IDs 100..106 are dedicated to the weekly reminder (one per ISO weekday).
  /// Using a fixed range makes cancellation deterministic.
  static const _baseId = 100;

  Future<void> initialise() async {
    if (_initialised) return;
    tz_data.initializeTimeZones();
    try {
      final localName = await FlutterTimezone.getLocalTimezone();
      tz.setLocalLocation(tz.getLocation(localName));
    } catch (_) {
      // Fallback to UTC — better than crashing.
    }
    const init = InitializationSettings(
      iOS: DarwinInitializationSettings(),
      android: AndroidInitializationSettings('@mipmap/ic_launcher'),
    );
    await _plugin.initialize(init);
    _initialised = true;
  }

  /// Show the OS-level permission prompt (or no-op if already granted).
  /// Returns true when the user grants the request.
  Future<bool> requestPermission() async {
    await initialise();
    try {
      final ios = _plugin.resolvePlatformSpecificImplementation<
          IOSFlutterLocalNotificationsPlugin>();
      if (ios != null) {
        final ok = await ios.requestPermissions(alert: true, sound: true);
        return ok ?? false;
      }
      final android = _plugin.resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin>();
      if (android != null) {
        final ok = await android.requestNotificationsPermission();
        return ok ?? true;
      }
    } catch (e) {
      if (kDebugMode) debugPrint('Notification permission error: $e');
    }
    return true;
  }

  /// Cancel any previously scheduled reminders and re-schedule based on
  /// [prefs]. Safe to call repeatedly — every save runs through this path.
  Future<void> reschedule(ReminderPrefs prefs) async {
    await initialise();
    await _plugin.cancelAll();
    if (!prefs.enabled || prefs.days.isEmpty) return;

    final parts = prefs.time.split(':');
    if (parts.length != 2) return;
    final hour = int.tryParse(parts[0]) ?? 18;
    final minute = int.tryParse(parts[1]) ?? 0;

    final details = NotificationDetails(
      iOS: const DarwinNotificationDetails(),
      android: AndroidNotificationDetails(
        'training_reminders',
        t('notif.channel_name'),
        channelDescription: t('notif.channel_desc'),
        importance: Importance.high,
        priority: Priority.high,
      ),
    );

    for (final iso in prefs.days) {
      if (iso < 1 || iso > 7) continue;
      try {
        await _plugin.zonedSchedule(
          _baseId + iso,
          t('notif.reminder_title'),
          t('notif.reminder_body'),
          _nextInstanceOfWeekday(iso, hour, minute),
          details,
          androidScheduleMode: AndroidScheduleMode.inexactAllowWhileIdle,
          // We anchor to the user's wall-clock — they expect 18:00 local
          // even if travelling. Some plugin versions still require the
          // legacy "absolute" interpretation flag.
          uiLocalNotificationDateInterpretation:
              UILocalNotificationDateInterpretation.absoluteTime,
          matchDateTimeComponents: DateTimeComponents.dayOfWeekAndTime,
        );
      } catch (e) {
        if (kDebugMode) debugPrint('Schedule failed for ISO $iso: $e');
      }
    }
  }

  /// Next occurrence of [isoWeekday] at the given local time. We compute in
  /// local TZ so the user's "18:00" means their 18:00 even when travelling.
  tz.TZDateTime _nextInstanceOfWeekday(
      int isoWeekday, int hour, int minute) {
    final now = tz.TZDateTime.now(tz.local);
    var scheduled = tz.TZDateTime(
        tz.local, now.year, now.month, now.day, hour, minute);
    while (scheduled.weekday != isoWeekday || !scheduled.isAfter(now)) {
      scheduled = scheduled.add(const Duration(days: 1));
    }
    return scheduled;
  }
}

File: apps/native-mobile/src/lib/notificationService.ts
// Ported 1:1 from apps/mobile/lib/services/notification_service.dart.
//
// Reimplemented on expo-notifications (managed Expo, SDK 57). The Flutter
// original wrapped `flutter_local_notifications` + the `timezone` package; here
// each weekly reminder becomes one `WEEKLY` calendar trigger. The public
// surface is preserved: initialise(), requestPermission(), reschedule(prefs).
// Re-entrant — every reschedule starts from a clean slate (cancelAll → schedule).
//
// PLATFORM SETUP: the first requestPermission() triggers the system prompt.
// expo-notifications needs no timezone bootstrap — the WEEKLY trigger fires in
// the device's local wall-clock time, matching the Flutter behaviour where the
// user's "18:00" means their 18:00.
import * as Notifications from "expo-notifications";
import { Platform } from "react-native";
import { t } from "../i18n";
import { ReminderPrefs } from "../models/auth";

// Android channel the reminders post to. Mirrors the Flutter original's
// 'training_reminders' AndroidNotificationDetails channel.
const ANDROID_CHANNEL_ID = "training_reminders";

class NotificationServiceImpl {
  private initialised = false;

  async initialise(): Promise<void> {
    if (this.initialised) return;
    // On Android, create the high-importance channel the reminders post to.
    // (iOS has no channels; the WEEKLY trigger carries the schedule directly.)
    if (Platform.OS === "android") {
      try {
        await Notifications.setNotificationChannelAsync(ANDROID_CHANNEL_ID, {
          name: t("notif.channel_name"),
          description: t("notif.channel_desc"),
          importance: Notifications.AndroidImportance.HIGH,
        });
      } catch {
        // Channel setup is best-effort — never block startup.
      }
    }
    this.initialised = true;
  }

  /// Show the OS-level permission prompt (or no-op if already granted).
  /// Returns true when the user grants (or the platform auto-grants) the request.
  async requestPermission(): Promise<boolean> {
    await this.initialise();
    try {
      const current = await Notifications.getPermissionsAsync();
      if (current.granted) return true;
      const req = await Notifications.requestPermissionsAsync({
        ios: { allowAlert: true, allowSound: true },
      });
      return req.granted;
    } catch (e) {
      if (__DEV__) console.warn("Notification permission error:", e);
    }
    return true;
  }

  /// Cancel any previously scheduled reminders and re-schedule based on [prefs].
  /// Safe to call repeatedly — every save runs through this path.
  async reschedule(prefs: ReminderPrefs): Promise<void> {
    await this.initialise();
    await Notifications.cancelAllScheduledNotificationsAsync();
    if (!prefs.enabled || prefs.days.length === 0) return;

    const parts = prefs.time.split(":");
    if (parts.length !== 2) return;
    const h = parseInt(parts[0], 10);
    const m = parseInt(parts[1], 10);
    const hour = Number.isNaN(h) ? 18 : h;
    const minute = Number.isNaN(m) ? 0 : m;

    for (const iso of prefs.days) {
      if (iso < 1 || iso > 7) continue;
      try {
        const trigger: Notifications.WeeklyTriggerInput = {
          type: Notifications.SchedulableTriggerInputTypes.WEEKLY,
          // Flutter used ISO weekdays (1=Mon … 7=Sun). expo-notifications'
          // WEEKLY weekday is 1=Sun … 7=Sat, so remap: (iso % 7) + 1
          // (ISO Mon→2, … ISO Sat→7, ISO Sun→1).
          weekday: (iso % 7) + 1,
          hour,
          minute,
        };
        if (Platform.OS === "android") trigger.channelId = ANDROID_CHANNEL_ID;
        await Notifications.scheduleNotificationAsync({
          content: {
            title: t("notif.reminder_title"),
            body: t("notif.reminder_body"),
          },
          trigger,
        });
      } catch (e) {
        if (__DEV__) console.warn(`Schedule failed for ISO ${iso}:`, e);
      }
    }
  }
}

/// Singleton — mirrors NotificationService.instance in the Dart original.
export const notificationService = new NotificationServiceImpl();
